### DISCUSSION TITLE
Feature Request: Support BS814A Touch Controller

### DISCUSSION THREAD

**Discussion by snookyly** [arendst/Tasmota#10195] [Ideas]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._
When trying to figure the setup of Teekar 3-gang wall switch I faced the issue that the touch switches utilizing the BS814A touch chip is not supported by tasmota.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [X] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): Teekar 3-gang wall switch
- [X] Tasmota binary firmware version number used: 9.1.0.2
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: Tuya-convert
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:

N/A
```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:

N/A
```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:

N/A
```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:

N/A
```

### TO REPRODUCE
_Steps to reproduce the behavior:_
Flash the switch with Tasmota

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

The three touch buttons should be detectable as switch inputs to Tasmota

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

As shown in the related manual (https://www.holtek.com/documents/10179/116711/BS81xA-xv160.pdf) here is the data transfer timing and the serial data stream format:

![image](https://user-images.githubusercontent.com/9679127/102700123-f3c94880-424a-11eb-9171-2ad32b17cda7.png)

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

Though being a software developer I am unfortunately not able to code the driver on my own due to limited skills in hardware development. Your help is highly appreciated. This request is related to https://github.com/arendst/Tasmota/discussions/10185.

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by vic42** [comment]
This is possibly going to be my new year's project. Waiting for the hardware to come in.

**Comment by vic42** [comment]
```
#ifdef USE_BS814A2
/*********************************************************************************************\
  THE PLAN [tm]:                      OUTCOME:
  ==============                      ========
  appear in a dropdown              
  select 2 pins (clk, data)                        
  poll data pin for low state
  generate high-speed clock (25kHz)
  poll data bits on clock plateau
  update data & report every 50 ms
  error & retry logic (if reqired)
  MQTT message & rules hook
```

**Comment by vic42** [comment]
Current driver clock timing:

![Spectacle wY9116](https://user-images.githubusercontent.com/3363362/103138905-dbd04980-46ce-11eb-86c4-9e32226f0aba.png)
```
#ifdef USE_BS814A2
/*********************************************************************************************\
  THE PLAN [tm]:                        OUTCOME:
  ==============                        ========
  appear in a dropdown                  (BS814 CLK, BS814 DAT)
  select 2 pins (clk, data)             DONE
  poll data pin for low state           DONE
  generate high-speed clock (25kHz)     DONE
  poll data bits on clock plateau       sample after rising clk edge
  update data & report every 50 ms      DONE
  error & retry logic (if reqired)      TODO
  MQTT message & rules hook             DONE
  Test on actual Hardware               TODO

  THE PROTOCOL [tm]:
  ==================
  see Holtek-Semicon BS814A-2 datasheet for details
\*********************************************************************************************/

```

**Comment by vic42** [comment]
First functional driver version
![Spectacle UV5542](https://user-images.githubusercontent.com/3363362/103836836-1b5d4500-5082-11eb-8458-81528a0cad7d.png)

**Comment by snookyly** [comment]
Successfully tested. Thanks for implementing.
Can be closed.
